core/src/main/java/nl/nn/adapterframework/filesystem/ExchangeFileSystemCache.java

```diff
@@ -0,0 +1,231 @@
+import org.apache.commons.lang3.StringUtils;
+import org.apache.logging.log4j.Logger;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Acts as a cache for EWS components.
+ * Stores Folder objects in memory for quick access.
+ *
+ * @author M63H114 - Laurens Mäkel.
+ */
+public class ExchangeFileSystemCache {
+	private Logger log = LogUtil.getLogger(this);
+
+	private final Map<String, FolderId> baseFolders = new HashMap<>(); // <Mailbox+FolderNaam, FolderId>
+	private final HashMap<String, Folder> folders = new HashMap<>(); // <Mailbox+FolderNaam, Folder>
+	private final List<String> mailboxesList = new ArrayList<>();
+
+	/**
+	 * Ensures that all Folder objects are present in cache.
+	 *
+	 * @param mailbox - Mailbox to ensure in cache.
+	 */
+	public void ensureMailboxIsRegistered(String mailbox, String baseFolderName, ExchangeService service) throws Exception {
+		if(!isMailboxRegistered(mailbox)){
+			registerMailbox(mailbox, baseFolderName, service);
+		}
+	}
+
+	/**
+	 * Retrieves Exchange Folder object from cache.
+	 *
+	 * @param mailbox - Mailbox to retrieve Folder from
+	 * @param folder - Folder name to look for
+	 *
+	 * @return Folder - The EWS Folder object that matches parameters.
+	 */
+	public Folder getFolder(String mailbox, String folder) throws IllegalStateException {
+		log.debug("Looking for folder ["+folder+"] in ["+mailbox+"]");
+
+		Folder result = folders.get(mailbox+folder);
+		if(result == null){
+			log.warn("Cannot find folder for mailbox ["+mailbox+"] and folder ["+folder+"].");
+			// throw new IllegalStateException("Cannot find folder for mailbox ["+mailbox+"] and folder ["+folder+"].");
+		}
+
+		return result;
+	}
+
+	public FolderId getBaseFolderId(String mailbox){
+		return baseFolders.get(mailbox);
+	}
+
+	/**
+	 * Checks if specified mailbox is already cached.
+	 *
+	 * @param mailbox - The name of the mailbox to check.
+	 *
+	 * @return boolean - Confirmation or denial.
+	 */
+	private boolean isMailboxRegistered(String mailbox){
+		return mailboxesList.contains(mailbox);
+	}
+
+
+	/**
+	 * Starts the process of caching all Folder objects for specified mailbox.
+	 *
+	 * @param mailbox - The name of the mailbox to cache.
+	 */
+	private synchronized void registerMailbox(String mailbox, String baseFolderName, ExchangeService service) throws Exception {
+		if(!isMailboxRegistered(mailbox)){
+			log.debug("Creating a local cache of folders for ["+mailbox+"].");
+
+			FolderId baseFolderId = getBaseFolderId(mailbox, baseFolderName, service);
+			baseFolders.put(mailbox, baseFolderId);
+
+			ArrayList<Folder> folders = findFolders(service, baseFolderId, Integer.MAX_VALUE);
+			for (Folder localFolder : folders) {
+				registerFolder(mailbox, localFolder);
+			}
+
+			mailboxesList.add(mailbox);
+		}
+	}
+
+	public FolderId getBaseFolderId(String emailAddress, String baseFolderName, ExchangeService service) throws FileSystemException {
```

Comment:
Deze method is vrijwel 1:1 verplaatst uit ExchangeFileSystem. Graag daar weer terug plaatsen, zodat duidelijk is wat de aanpassingen zijn. Hij heeft voor zover ik kan zien geen afhankelijkheden met ExchangeFileSystemCache